### ISSUE TITLE
Can't build a docker container. Version v9.0.1 

### ISSUE THREAD

**Issue by MrOxMasTer** [pnpm/pnpm#7938]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

v9.0.1

### Which area(s) of pnpm are affected? (leave empty if unsure)

CLI, Lockfile

### Reproduction steps

when assembling the container

### Describe the Bug

![image](https://github.com/pnpm/pnpm/assets/59291123/0e75dc36-e774-43e7-a589-5a177e9b4317)
![image](https://github.com/pnpm/pnpm/assets/59291123/cb77492d-af17-4f7c-8c51-bfb33c4c3d98)
![image](https://github.com/pnpm/pnpm/assets/59291123/1d06c651-ce26-4680-b92f-ec3f68c41cbe)

container breaks

### Expected Behavior

the container was being assembled

### Which Node.js version are you using?

Global: 20.11.1 / In container: node:20-alpine

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [X] Linux

**Comment by MrOxMasTer** [comment]
> Could you attach the package.json with the dependencies?

```
{
  "name": "shopco",
  "version": "0.1.0",
  "private": true,
  "type": "module",
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint",
    "migration:generate": "drizzle-kit generate:pg",
    "migration:push": "drizzle-kit push:pg",
    "studio": "drizzle-kit studio --host 127.0.0.1",
    "build:icons": "npx tsx ./build-icons.mts",
    "sly": "pnpm dlx @sly-cli/sly add",
    "prepare": "husky"
  },
  "dependencies": {
    "@auth/drizzle-adapter": "^0.8.0",
    "@hookform/resolvers": "^3.3.4",
    "@radix-ui/react-dialog": "^1.0.5",
    "@radix-ui/react-slot": "^1.0.2",
    "@t3-oss/env-nextjs": "^0.9.2",
    "argon2": "^0.40.1",
    "class-variance-authority": "^0.7.0",
    "clsx": "^2.1.0",
    "drizzle-orm": "^0.30.2",
    "drizzle-zod": "^0.5.1",
    "embla-carousel-autoplay": "8.0.0-rc17",
    "embla-carousel-react": "8.0.0-rc17",
    "ky": "^1.2.2",
    "nanoid": "^5.0.6",
    "next": "14.2.1",
    "next-auth": "5.0.0-beta.16",
    "next-themes": "^0.3.0",
    "pg": "^8.11.3",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-focus-lock": "^2.11.2",
    "react-hook-form": "^7.51.2",
    "redis": "^4.6.13",
    "server-only": "^0.0.1",
    "sharp": "^0.33.2",
    "sonner": "^1.4.3",
    "tailwind-merge": "^2.2.1",
    "tailwindcss-animate": "^1.0.7",
    "zod": "^3.22.4",
    "zustand": "^4.5.2"
  },
  "devDependencies": {
    "@conarti/eslint-plugin-feature-sliced": "^1.0.5",
    "@sly-cli/sly": "^1.10.0",
    "@types/fs-extra": "^11.0.4",
    "@types/node": "^20.11.28",
    "@types/pg": "^8.11.2",
    "@types/react": "^18.2.66",
    "@types/react-dom": "^18.2.22",
    "@typescript-eslint/eslint-plugin": "^6.21.0",
    "@typescript-eslint/parser": "^6.21.0",
    "autoprefixer": "^10.4.18",
    "cssnano": "^6.1.0",
    "drizzle-kit": "^0.20.14",
    "eslint": "^8.57.0",
    "eslint-config-next": "14.1.3",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-drizzle": "^0.2.3",
    "eslint-plugin-import": "^2.29.1",
    "eslint-plugin-promise": "^6.1.1",
    "eslint-plugin-tailwindcss": "^3.14.3",
    "fs-extra": "^11.2.0",
    "glob": "^10.3.10",
    "husky": "^9.0.11",
    "jiti": "^1.21.0",
    "lucide-react": "^0.303.0",
    "node-html-parser": "^6.1.12",
    "postcss": "^8.4.35",
    "postcss-flexbugs-fixes": "^5.0.2",
    "postcss-preset-env": "^9.4.0",
    "prettier": "^3.2.5",
    "prettier-plugin-tailwindcss": "^0.5.12",
    "react-devtools": "^5.0.2",
    "stylelint": "^16.3.0",
    "stylelint-config-standard": "^36.0.0",
    "stylelint-order": "^6.0.4",
    "stylelint-selector-bem-pattern": "^4.0.0",
    "svgo": "^3.2.0",
    "tailwindcss": "^3.4.1",
    "tailwindcss-debug-screens": "^2.2.1",
    "typescript": "^5.4.2"
  },
  "packageManager": "pnpm@9.0.1+sha256.46d50ee2afecb42b185ebbd662dc7bdd52ef5be56bf035bb615cab81a75345df"
}

```

**Comment by zkochan** [comment]
seems to work on macOS

Did it work with pnpm v8?

**Comment by MrOxMasTer** [comment]
> seems to work on macOS
> 
> Did it work with pnpm v8?

Yeah, I was literally building a container today. Then I went back to it, tried again and got an error. And I see that there's a new update to pnpm.

**Comment by MrOxMasTer** [comment]
Inside the container it downloads the new version to me, there when the build happens it returns the lines from the console and there was a line about pnpm 9.0.1.

**Comment by MrOxMasTer** [comment]
![image](https://github.com/pnpm/pnpm/assets/59291123/e748c8f1-7ce6-4b3a-8676-bdf08fc1dbc1)
Well maybe it might not be a pnpm problem. Mb node.js or docker, although just today everything was fine this morning

**Comment by MrOxMasTer** [comment]
> seems to work on macOS
> 
> Did it work with pnpm v8?

The problem is in the version. And even before version 9 it was there. I installed 8.15.7 and it didn't build on my pc either.
I installed version 8.15.4 and everything is assembled

**Comment by MrOxMasTer** [comment]
Tried installing the 9.0.1 version and it worked.

![image](https://github.com/pnpm/pnpm/assets/59291123/06e004f8-bd5d-4436-8810-9bd966f179ef)
![image](https://github.com/pnpm/pnpm/assets/59291123/08f5fc56-a7af-4bc3-be59-4de45c3f3c3c)

**Comment by MrOxMasTer** [comment]
I don't know what happened. Either because I reinstalled docker and rolled it back to version 4.28, and before that it was 4.29, which I recently updated. Or because pnpm 9.0.0 broke the build and the old version fixed it somehow

**Comment by MrOxMasTer** [comment]
> Did it work with pnpm v8?

It was probably a bug in docker 4.29. Previously all the above successful builds were done on 4.28. Now when I upgraded to 4.29 again when building, it showed me the error I had for a moment, but then it disappeared and the build ran successfully

![image](https://github.com/pnpm/pnpm/assets/59291123/f2aaa0aa-4aa0-4a9a-aefb-7354a7670aa4)

**Comment by divmgl** [comment]
We're having similar issues building with pnpm@9. In Docker we're getting:

```
------
 > [server dependencies 5/5] RUN pnpm i --frozen-lockfile --prod:
0.384 Scope: all 3 workspace projects
0.485  ERR_PNPM_LOCKFILE_CONFIG_MISMATCH  Cannot proceed with the frozen installation. The current "pnpmfileChecksum" configuration doesn't match the value found in the lockfile
0.485
0.485 Update your lockfile using "pnpm install --no-frozen-lockfile"
------
failed to solve: process "/bin/sh -c pnpm i --frozen-lockfile --prod" did not complete successfully: exit code: 1
```

This is breaking CI and local machines for us.

**Comment by MrOxMasTer** [comment]
> We're having similar issues building with pnpm@9. In Docker we're getting:
> 
> ```
> ------
>  > [server dependencies 5/5] RUN pnpm i --frozen-lockfile --prod:
> 0.384 Scope: all 3 workspace projects
> 0.485  ERR_PNPM_LOCKFILE_CONFIG_MISMATCH  Cannot proceed with the frozen installation. The current "pnpmfileChecksum" configuration doesn't match the value found in the lockfile
> 0.485
> 0.485 Update your lockfile using "pnpm install --no-frozen-lockfile"
> ------
> failed to solve: process "/bin/sh -c pnpm i --frozen-lockfile --prod" did not complete successfully: exit code: 1
> ```
> 
> This is breaking CI and local machines for us. We're also using corepack to install pnpm.

This was fixed in 9.0.1, I think.

**Comment by divmgl** [comment]
Still broken in 9.0.1 for us. Works locally but getting this error in the Docker image. Confirmed the version is right and the pnpm-lock.yaml file is correct.

**Comment by elrumordelaluz** [comment]
Same here

**Comment by zkochan** [comment]
>  ERR_PNPM_LOCKFILE_CONFIG_MISMATCH  Cannot proceed with the frozen installation. The current "pnpmfileChecksum" configuration doesn't match the value found in the lockfile

Update the lockfile.

**Comment by milanpajovic** [comment]
Same issue, AWS app runner is broken from yesterday

**Comment by zkochan** [comment]
> Same issue, AWS app runner is broken from yesterday
Error:  ERR_PNPM_BAD_PM_VERSION  This project is configured to use v8.10.4 of pnpm. Your current pnpm is v9.0.1

It is not broken. You have specified the `packageManager` field in `package.json`. Use pnpm v8.10.4 or update the field in `packageManager`. Or use corepack to install pnpm.

**Comment by milanpajovic** [comment]
> > Same issue, AWS app runner is broken from yesterday
> > Error:  ERR_PNPM_BAD_PM_VERSION  This project is configured to use v8.10.4 of pnpm. Your current pnpm is v9.0.1
> 
> It is not broken. You have specified the `packageManager` field in `package.json`. Use pnpm v8.10.4 or update the field in `packageManager`. Or use corepack to install pnpm.
This is how my package.json looks like:


  "packageManager": "pnpm@8.10.4"

**Comment by zkochan** [comment]
So either use that version of pnpm or change the field to the version that you want to run. There's no issue on pnpm's end.

**Comment by divmgl** [comment]
Hey guys, I want to be clear here. We've identified that we're using 9.0.1 to both generate the lock file and we're also using 9.0.1 in the Docker image. This is still an ongoing issue.

**Comment by divmgl** [comment]
To anyone who is having issues with `pnpm` in Render.com, add an environment variable named `SKIP_INSTALL_DEPS` and set it to `true` and then use this command to install `pnpm@8` and unblock your builds:

```
mkdir -p bin; corepack enable --install-directory bin; corepack prepare pnpm@8.14.2 --activate; export PATH="$PWD/bin:$PATH"; pnpm i --frozen-lockfile; pnpm app build
```

Source: https://docs.render.com/static-sites#dependency-installation

**Comment by divmgl** [comment]
Shout out to @zkochan as he found the solution to our problem: **make sure to check that `.pnpmfile.cjs` is included in your Docker builds.**

**Comment by AntonOfTheWoods** [comment]
@divmgl , are you saying that that file is now obligatory if we want to use `pnpm` in docker for building? It definitely isn't required outside of docker...

**Comment by zkochan** [comment]
@AntonOfTheWoods in the past we had a problem, users modified the `.pnpmfile.cjs` file but the lockfile was not updated. We have added the checksum of the pnpmfile to the lockfile. So, now if the checksum changes pnpm updates the lockfile. If the pnpmfile is missing but the lockfile was created by a pnpmfile, pnpm will recreated the lockfile without pnpmfile OR fail if install is executed with `--frozen-lockfile`.

**Comment by nathan-vm** [comment]
Hello everyone, I have the same problem with `pnpm`, every time I use my `docker compose up`, pnpm adds the `packageManager` variable to package.json (on compose I have the volume mapped). This means that if `pnpm` updates, the message appears:

```ERR_PNPM_LOCKFILE_CONFIG_MISMATCH  Cannot proceed with the frozen installation. The current "pnpmfileChecksum" configuration doesn't match the value found in the lockfile
```

I didn't find in the documentation how to prevent the `packageManager` key from being added to `package.json`, and so every time this key is added I need to be careful not to add it to my git repositor.

![Captura de Tela 2024-05-31 às 14 40 10](https://github.com/pnpm/pnpm/assets/54302847/598d8db0-c75f-4e95-9a1e-bc1e7e9152b8)

![image](https://github.com/pnpm/pnpm/assets/54302847/e43464cf-a895-4f80-8037-8d6f6cf5d9a0)

**Comment by zkochan** [comment]
Closing for lack of a reproduction on a supported version.

There are only screenshots here. Please open a new issue with the Dockerfile and the output if a current pnpm still fails in your image.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
